tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,301 @@
+                # so if you have a resnet and you specifically assign the first resnet layer
+                # to have lr_mult of 0.01 and the resnet model to have lr_mult of 0.1, all
+                # resnet layers except the first should get lr_mult of 0.1 and the first
+                # keeps its lr_mult of 0.01
+                DiscriminativeModelManager._assign_lr_mult(sublayer, mult)
+
+                # recursively iterate through the nested layers
+                for (
+                    nested_sublayer
+                ) in DiscriminativeModelManager._recursively_assign_sublayer_lr_mult(
+                    sublayer
+                ):
+                    yield nested_sublayer
+        else:
+            yield layer
+
+    @staticmethod
+    def _apply_lr_mult_to_var(layer):
+        """Helper method to apply the lr mult to the trainable variables of a layer
+        """
+        lr_mult = DiscriminativeModelManager._get_lr_mult(layer)
+        for var in layer.trainable_variables:
+            var.lr_mult = lr_mult
+            # the lr_mult behaves as a hyper parameter and not a variable. it will not be a tensor
+            # there's not benefit in setting the lr_mult as a variable because it does not interact with tensors
+
+    @staticmethod
+    def _check_for_lr_mult(layer, verbose=True, propagate=True):
+        """Identify which layers have an LR mult not equal to 1
+        """
+
+        layers_with_lr_mult = []
+
+        for (
+            sub_layer
+        ) in DiscriminativeModelManager._recursively_assign_sublayer_lr_mult(layer):
+            lr_mult = DiscriminativeModelManager._get_lr_mult(sub_layer)
+            if lr_mult != 1.0:
+                layers_with_lr_mult.append(sub_layer)
+                if verbose:
+                    logging.info("layer %s lr_mult : %f" % (sub_layer.name, lr_mult))
+
+        return layers_with_lr_mult
+
+    @staticmethod
+    def _compute_params(var_list):
+        """helps compute params to provide a summary that aligns with model.summary()
+        """
+        return np.sum([np.prod(list(var.shape)) for var in var_list])
+
+    @staticmethod
+    def _prepare_model(model, verbose=True):
+        """Prepares a model for disc training
+        """
+
+        layers_with_lr_mult = DiscriminativeModelManager._check_for_lr_mult(
+            model, verbose=verbose
+        )
+        if len(layers_with_lr_mult) == 0:
+
+            logging.warning(
+                """No Layer has been assigned an lr_mult attribute != 1.0
+                Discriminative Layer Training will apply the same learning rate to all layers
+                It will perform as if you did not use Discriminative Layer Training
+                """
+            )
+
+        for layer in DiscriminativeModelManager._recursively_assign_sublayer_lr_mult(
+            model
+        ):
+            DiscriminativeModelManager._apply_lr_mult_to_var(layer)
```

Comment:
Can you please explain the logic of doing both:
```
layer.lr_mult = lr_mult
```
and 
```
layer.variable.lr_mult = lr_mult
``` 

Shouldn't it be one or the other?

Reply:
I've added an additional comment to explain the reasoning. Might not be the best reasoning tbh but it makes me feel safe.

Reply:
Makes perfect sense, thanks for the clarification.